qualang_tools/optimal_weights/StateDiscriminator.py

```diff
@@ -0,0 +1,256 @@
+            period_ns = int(1 / rr_freq * 1e9)
+            hann = signal.hann(period_ns * 2, sym=True)
+            hann = hann / np.sum(hann)
+            traces = np.array([np.convolve(traces[i, :], hann, 'same') for i in range(self.num_of_states)])
+        return traces
+
+    @staticmethod
+    def _quantize_traces(traces):
+        weights = []
+        for i in range(traces.shape[0]):
+            weights.append(np.average(np.reshape(traces[i, :], (-1, 4)), axis=1))
+        return np.array(weights)
+
+    def _execute_and_fetch(self, program, **execute_args):
+        qm = self.qmm.open_qm(self.config)
+        job = qm.execute(program, duration_limit=0, data_limit=0, **execute_args)
+        res_handles = job.result_handles
+        res_handles.wait_for_all_values()
+        I_res = res_handles.get("I").fetch_all()['value']
+        I_res = np.concatenate([I_res[0::2], I_res[1::2]])
+        Q_res = res_handles.get("Q").fetch_all()['value']
+        Q_res = np.concatenate([Q_res[0::2], Q_res[1::2]])
+
+        if I_res.shape != Q_res.shape:
+            raise RuntimeError("")
+
+        ts = res_handles.get("adc1").fetch_all()['value']['timestamp']
+        ts = np.concatenate([ts[0::2], ts[1::2]])
+        in1 = res_handles.get("adc1").fetch_all()['value']['value']
+        in1 = np.concatenate([in1[0::2], in1[1::2]])
+        in2 = res_handles.get("adc2").fetch_all()['value']
+        in2 = np.concatenate([in2[0::2], in2[1::2]])
+        if not self.lsb:
+            x = in1 + 1j * in2
+        else:
+            x = in1 - 1j * in2
+        return I_res, Q_res, ts, x
+
+    def train(self, program, use_hann_filter=True, plot=False, correction_method='robust', **execute_args):
+        """
+        The train procedure is used to calibrate the optimal weights and bias for each state. A file with the optimal
+        parameters is generated during training, and it is used during the subsequent measure_state procedure.
+        A training program must be provided in the constructor.
+        :param program: a training program. A program that generates training sets. The program should generate equal
+        number of training sets for each one of the states. Collection of training sets is achieved by first preparing
+        the qubit in one of the states, and then measure the readout resonator element. The measure command must include
+        streaming of the raw data (the tag must be called "adc") and the final complex demodulation results (which is
+        constructed from 4 real demodulations) must be saved under the tags "I" and "Q". E.g:
+
+            measure("readout", "rr", "adc", demod.full("integW_cos", I1, "out1"),
+                                            demod.full("integW_sin", Q1, "out1"),
+                                            demod.full("integW_cos", I2, "out2"),
+                                            demod.full("integW_sin", Q2, "out2"))
+            assign(I, I1 + Q2)
```

Comment:
Need to change to dual demod

Reply:
yes, I will update the comment and also the config in the example as well